manager/allocator/networkallocator/networkallocator.go

```diff
@@ -108,12 +111,18 @@ type DriverValidator interface {
 	ValidateIPAMDriver(*api.Driver) error
 }
 
+// NetworkStateUpdater is an interface for updating the network state in the store.
+type NetworkStateUpdater interface {
+	UpdateNetworkState(*api.Network) error
+}
+
 // Provider provides network allocation functionality.
 type Provider interface {
 	DriverValidator
 	PredefinedNetworks() []PredefinedNetworkData
 	SetDefaultVXLANUDPPort(uint32) error
 	NewAllocator(*Config) (NetworkAllocator, error)
+	NetworkStateUpdater() NetworkStateUpdater
```

Comment:
I know how these interfaces are implemented on the Moby side. `cnmallocator.cnmNetworkAllocator` holds the IPAM state, is not a singleton, and is not an implementation of Provider. This method could only be implemented on `cnmallocator.Provider` using horrible hacks. There has to be a better way.